src/Development/IDE/Core/Rules.hs

```diff
@@ -570,7 +582,32 @@ getDocMapRule =
       dkMap <- liftIO $ mkDocMap hsc parsedDeps rf tc
       return ([],Just dkMap)
 
--- Typechecks a module.
+persistentHieFileRule :: Rules ()
+persistentHieFileRule = addPersistentRule GetHieAst $ \file -> runMaybeT $ do
+  res <- readHieFileForSrcFromDisk file
+  let refmap = generateReferencesMap . getAsts . hie_asts $ res
+  pure $ HAR (hie_module res) (hie_asts res) refmap (HieFromDisk res)
+
+readHieFileForSrcFromDisk :: NormalizedFilePath -> MaybeT IdeAction HieFile
+readHieFileForSrcFromDisk file = do
+  db <- asks hiedb
+  log <- asks $ L.logInfo . logger
+  row <- MaybeT $ liftIO $ HieDb.lookupHieFileFromSource db $ fromNormalizedFilePath file
+  let hie_loc = HieDb.hieModuleHieFile row
+  liftIO $ log $ "LOADING HIE FILE :" <> T.pack (show file)
+  readHieFileFromDisk hie_loc
```

Comment:
I am trying to understand what this buys us. Ghcide already reuses `.hie` files from the cache dir at startup if they are up-to-date, so in what cases does this make a difference?

Reply:
> Ghcide already reuses .hie files from the cache dir at startup if they are up-to-date

It doesn't, at least not for things like hover/definition etc.

With the persistent rule stuff we get:

- Instant availability of select IDE features (hover/definition...) on startup, even before `cabal configure` etc. finish
- If the project doesn't compile on startup, we can still use the above features from the stale `.hie` file.

Reply:
That's awesome. 

But surely ghcide will block until the cradle is loaded, as it usually happens on a Notification?

Reply:
> But surely ghcide will block until the cradle is loaded, as it usually happens on a Notification?

Ah, this is indeed the behavior in vscode. I didn't notice because my client sends the request immediately after initialize if the server isn't already running, so I was getting instant responses most of the time.

I think we can serialize notifications with respect to each other, but handle requests concurrently without blocking?

Reply:
We decided to process notifications synchronously to ensure consistency after seeing test failures showing real inconsistency issues around e.g. completions. I don't see how to avoid such inconsistencies if we handle requests concurrently with notifications.

Reply:
https://github.com/mpickering/hie-bios/pull/265 will make startup much faster when it lands, so we will get most of the benefits of quick startup even without async notifications. Do we still want to keep this mechanism to provide code navigation when the project doesn't compile on startup, or should I just remove it?

If we are keeping it, I think it needs a re-design. I'm thinking another variant of `define` in which we can return a stale value if the regular codepath fails.

Reply:
I think you've answered yourself! Get rid of this version and leave the redesign for future work

Reply:
I think that my assumption that ghcide blocks on a FileOpened Notification might be wrong. 
The handler calls `restartShakeSession` but doesn't wait on the results, so it might not be actually blocking. 
The open telemetry stuff will make it very easy to find out whether it blocks or not.